enhancements/20220413-zero-downtime-upgrade.md

```diff
@@ -0,0 +1,204 @@
+# Zero Downtime Upgrade
+
+## Summary
+
+Users might run mission-critical workloads in Harvester clusters. The zero-downtime upgrade support can allow users to upgrade their Harvestr clusters without interrupting services.
+
+### Related Issues
+
+- https://github.com/harvester/harvester/issues/1022
+
+## Motivation
+
+### Goals
+
+The user can upgrade a Harvester cluster to a newer version from GUI or API. During the upgrade, there will be no downtime for the VMs in the Harvester.
+
+### Non-goals
+
+- Upgrade OSes and software stacks inside VMs.
+
+## Proposal
+
+We'll implement an upgrade controller to manage `upgrade` CRs and handle upgrades.
+
+### User Stories
+
+#### 1. Multiple-nodes Harvester deployments (node count >= 3)
+
+The user can select a new Harvester version from GUI to perform cluster upgrades. During an upgrade, VMs are live-migrated between nodes to avoid guest service downtime. Harvester API endpoints, components, and VIP are also available during the upgrade.
+
+#### 2. 2-nodes Harvester deployments
+
+The user can select a new Harvester version from GUI to perform cluster upgrades. During an upgrade, VMs are live-migrated between nodes to avoid guest service downtime. Because there is only one server node, upgrading the server node brings downtime to Harvester API endpoints, components, VIP, and VMs' connectivity (depends on the network configuration).
+
+#### 3. Single node Harvester deployments
+
+The user can select a new Harvester version from GUI to perform cluster upgrades. VMs are shut down before an upgrade in this case.
+
+### User Experience In Detail
+
+GUI:
+- When there is a new upgradable version, the Harvester GUI will display an "upgrade" button to indicate there are new versions.
+- The user can click the "upgrade" button to select a new version from a drop-list to start the upgrade.
+- During an upgrade, a dialog displays the upgrade progress.
+
+API:
+- A new `Version` CR is automatically created when there is a new upgradable version.
+- The user can create an `Upgrade` CR to start an upgrade.
+
+### API changes
+
+No. The first implementation is included in the GA release 1.0.0.
+
+## Design
+
+### Implementation Overview
+
+#### Version management
+
+- When a new Harvester version is released, developers need to edit the response file in the [upgrade responder](https://github.com/harvester/upgrade-responder) to include the new [Version](https://github.com/harvester/upgrade-responder/blob/032365e51c204dfc4aa8793410f5e6a0114c276e/upgraderesponder/service.go#L60). Developers also need to specify the [`MinUpgradableVersion`](https://github.com/harvester/upgrade-responder/blob/032365e51c204dfc4aa8793410f5e6a0114c276e/upgraderesponder/service.go#L63) of the new version.
+- Harvester backend queries new versions from the upgrade responder hourly.
```

Comment:
If Harvester is deployed in an air-gapped environment, can users still receive the latest available upgrades? Or they should manually add versions CR?

Reply:
> can users still receive the latest available upgrades?

No, from GUI, we have a hidden page to let users upload the ISO and the frontend will create the CR.
Of course, it's also possible to create CR manually.

Reply:
@johnliu55tw I added the air-gapped scenario: https://github.com/harvester/harvester/pull/2146/commits/9a68a3a0a9754caa1bd51c5907783a975f8e3e63#diff-4024911e730bfb92019ae2cf3868136ba568e013b9d1d41fc09bdf27b86c6b37R41-R47